metricbeat/module/kubernetes/util/kubernetes.go

```diff
@@ -129,155 +149,402 @@ func getResource(resourceName string) kubernetes.Resource {
+			jobWatcher = (*watcher).watcher
+		}
+
+		metaGen = metadata.GetPodMetaGen(commonConfig, (*resWatcher).watcher, nodeWatcher, namespaceWatcher, replicaSetWatcher,
+			jobWatcher, config.AddResourceMetadata)
+		return metaGen, nil
+	} else if resourceName == ServiceResource {
+		namespaceWatcher := resourceWatchers.watchersMap[NamespaceResource]
+		if namespaceWatcher == nil {
+			return nil, fmt.Errorf("could not create the metadata generator, as the watcher for namespace does not exist")
+		}
+		namespaceMeta := metadata.NewNamespaceMetadataGenerator(config.AddResourceMetadata.Namespace,
+			(*namespaceWatcher).watcher.Store(), client)
+		metaGen = metadata.NewServiceMetadataGenerator(commonConfig, (*resWatcher).watcher.Store(),
+			namespaceMeta, client)
+		return metaGen, nil
+	}
+
+	// Should never reach this part, as this function is only for service or pod resources
+	return metaGen, fmt.Errorf("failed to create a metadata generator for resource %s", resourceName)
+}
+
 func NewResourceMetadataEnricher(
 	base mb.BaseMetricSet,
 	resourceName string,
 	metricsRepo *MetricsRepo,
+	resourceWatchers *Watchers,
 	nodeScope bool) Enricher {
 
-	var replicaSetWatcher, jobWatcher kubernetes.Watcher
+	log := logp.NewLogger(selector)
 
 	config, err := GetValidatedConfig(base)
 	if err != nil {
-		logp.Info("Kubernetes metricset enriching is disabled")
+		log.Info("Kubernetes metricset enriching is disabled")
 		return &nilEnricher{}
 	}
 
-	res := getResource(resourceName)
-	if res == nil {
+	// This type of config is needed for the metadata generator
+	commonMetaConfig := metadata.Config{}
+	if err := base.Module().UnpackConfig(&commonMetaConfig); err != nil {
+		log.Errorf("Error initializing Kubernetes metadata enricher: %s", err)
 		return &nilEnricher{}
 	}
+	commonConfig, _ := conf.NewConfigFrom(&commonMetaConfig)
 
 	client, err := kubernetes.GetKubernetesClient(config.KubeConfig, config.KubeClientOptions)
 	if err != nil {
-		logp.Err("Error creating Kubernetes client: %s", err)
+		log.Errorf("Error creating Kubernetes client: %s", err)
 		return &nilEnricher{}
 	}
 
-	watcher, nodeWatcher, namespaceWatcher := getResourceMetadataWatchers(config, res, client, nodeScope)
-
-	if watcher == nil {
+	err = startAllWatchers(client, resourceName, nodeScope, config, log, resourceWatchers)
+	if err != nil {
+		log.Errorf("Error starting the watchers: %s", err)
 		return &nilEnricher{}
 	}
 
-	// commonMetaConfig stores the metadata configuration of the resource itself
-	commonMetaConfig := metadata.Config{}
-	if err := base.Module().UnpackConfig(&commonMetaConfig); err != nil {
-		logp.Err("Error initializing Kubernetes metadata enricher: %s", err)
+	var specificMetaGen metadata.MetaGen
+	var generalMetaGen *metadata.Resource
+	if resourceName == ServiceResource || resourceName == PodResource {
+		specificMetaGen, err = createMetadataGenSpecific(client, commonConfig, config, resourceName, resourceWatchers)
```

Comment:
Only `config.AddResourceMetadata` is needed from the whole config. Same for `createMetadataGen`

Reply:
I mean that you do not need to pass the whole config in the `createMetadataGenSpecific` function. Just pass the `config.AddResourceMetadata`